flambda2: Provide an interface for rewriting of types

bclement-ocp opened the pull request:
Only the part of #5043 that is in the typing env. This introduces a new API that is currently unused, but will be used in #5043.

lthls commented on middle_end/flambda2/types/traversals.ml:
```diff
@@ -0,0 +1,1335 @@
+    let other_closures, acc =
+      match other_closures with
+      | Bottom -> Or_bottom.Bottom, acc
+      | Ok { maps_to = _; env_extension = _; index = _ } -> assert false
```
I think this deserves a comment (i.e. we should remove the `other_closures` field completely, as it is always Bottom, but not in this PR).

lthls commented on middle_end/flambda2/types/traversals.ml:
```diff
@@ -0,0 +1,1335 @@
+    in
+    TG.Row_like_for_closures.create_raw ~known_closures ~other_closures, acc
+
+  and rewrite_closures_entry ~this_function_slot:_ env acc metadata
```
It's a bit weird to have a parameter that is never used. Is it expected that it would become used later ? Or a leftover from a previous design ?

bclement-ocp replied:
I removed this argument.

lthls reviewed (approved):
I've gone through all the changes. I'm approving the PR as I believe it is computing the right thing and doesn't have unexpected side effects. If it is needed as part of an urgent fix it should be fine.

When we have time for properly working on this, here are the things that caught my eye:
- The entry points (`rewrite` and `rewrite_env_extension_with_extra_variables`) do not share enough code to my taste
- The `rewrite` entry point takes as parameter a map of live variables that is expected to always be empty
- Not all recursive calls to `rewrite_arbitrary_type` change the metadata. This works for the reaper because all the cases where the reaper has metadata are covered, but I would be in favour of enhancing the signature of `Traversals.Make` with metadata transformers for all type accesses (i.e. array lengths, boxed number contents, ...) to better reflect the actual traversal.
- There are still a few `assert false` left that should be either better commented or transformed to `Misc.fatal_error`

bclement-ocp commented on middle_end/flambda2/types/traversals.ml:
```diff
@@ -0,0 +1,1335 @@
+
+  let empty = { aliases_of_names = Name.Map.empty; names_to_process = [] }
+
+  let get_canonical_with ({ aliases_of_names; names_to_process } as u) _env
```
Done; `get_canonical_with` returns a pair now.

bclement-ocp commented:
Thank you for the review! I have addressed some of your overall remarks that were easy to resolve, and added a CR with the more complex one (adding accessors for all the type accesses).

As you note, since this is going to only be needed for the reaper at the moment, I think it is enough to merge as is, and we will include a cleanup of this module during the review of the reaper.
